onnx/shape_inference/implementation.cc

```diff
@@ -161,5 +180,26 @@ void InferShapes(
   }
 }
 
+void InferShapeForFunctionNode(
+    const NodeProto& node,
+    const FunctionProto& func,
+    const ISchemaRegistry* schema_registry,
+    int domain_version,
+    InferenceContext& ctx) {
+  // Create a temproary graphproto to hold the expanded subgraph
+  GraphProto g;
+  // To Generate unique internal tensor names
+  // while preserving node's input/output names
+  FunctionExpandHelper(node, func, g);
+  for (auto& n : g.node()) {
+    const auto schema =
+        schema_registry->GetSchema(n.op_type(), domain_version, n.domain());
+    if (!schema) {
+      return;
+    }
+    schema->GetTypeAndShapeInferenceFunction()(ctx);
```

Comment:
this logic is wrong. You should construct you own InferenceContext locally to do the sub-graph inference, and finally set the output type in ctx with the sub-graph's output.